File: src/chess_club/glicko2.py
import math
import chess_club.config as config

# Compact Glicko-2 implementation for two-player updates.
# Constants
TAU = 0.5
EPSILON = 1e-6
Q = math.log(10) / 400

def _g(phi):
    return 1 / math.sqrt(1 + (3 * (phi ** 2)) / (math.pi ** 2))

def _E(mu, mu_j, phi_j):
    return 1 / (1 + math.exp(-_g(phi_j) * (mu - mu_j)))

def _to_mu(r):
    return (r - 1500.0) / 173.7178

def _to_phi(rd):
    return rd / 173.7178

def _to_rating(mu):
    return mu * 173.7178 + 1500.0

def _to_rd(phi):
    return phi * 173.7178


def inflate_rd(rd: float, days: float) -> float:
    """Inflate RD according to inactivity days using config.G2_RD_INCREASE_PER_DAY.
    Caps the inflated RD to config.G2_DEFAULT_RD.
    """
    try:
        if not config.G2_RD_INCREASE_PER_DAY:
            return rd

        days = max(0.0, days)
        if days == 0:
            return rd

        c = config.G2_RD_INCREASE_PER_DAY
        rd_star = math.sqrt(rd * rd + (c * c) * days)
        return min(rd_star, config.G2_DEFAULT_RD)

    except (TypeError, ValueError):
        return rd


def _f(x, delta, phi, v, a, tau):
    ex = math.exp(x)
    num = ex * (delta * delta - phi * phi - v - ex)
    den = 2 * (phi * phi + v + ex) ** 2
    return (num / den) - ((x - a) / (tau * tau))

def glicko2_update(r, rd, vol, opp_r, opp_rd, opp_vol, score, tau=TAU, days: float = 0.0):
    # Convert to Glicko-2 scale
    mu = _to_mu(r)
    # increase RD according to inactivity/time using config.G2_RD_INCREASE_PER_DAY
    # rd is in rating points; variance should grow linearly with time so
    # rd_star = sqrt(rd^2 + c^2 * days)
    if days and config.G2_RD_INCREASE_PER_DAY:
        c = config.G2_RD_INCREASE_PER_DAY
        rd_star = math.sqrt(rd * rd + (c * c) * days)
    else:
        rd_star = rd

    phi = _to_phi(rd_star)
    mu_j = _to_mu(opp_r)
    phi_j = _to_phi(opp_rd)

    g = _g(phi_j)
    E = _E(mu, mu_j, phi_j)
    v = 1 / (g * g * E * (1 - E))
    delta = v * g * (score - E)

    a = math.log(vol * vol)
    A = a
    B = None
    if delta * delta > (phi * phi + v):
        B = math.log(delta * delta - phi * phi - v)
    else:
        k = 1
        while _f(a - k * tau, delta, phi, v, a, tau) < 0:
            k += 1
        B = a - k * tau

    fA = _f(A, delta, phi, v, a, tau)
    fB = _f(B, delta, phi, v, a, tau)

    # Binary search for sigma with iteration limit to avoid infinite loops
    max_iters = 60
    iters = 0
    while abs(B - A) > EPSILON and iters < max_iters:
        C = A + (A - B) * fA / (fB - fA)
        fC = _f(C, delta, phi, v, a, tau)
        if fC * fB < 0:
            A = B
            fA = fB
            B = C
            fB = fC
        else:
            fA = fA / 2.0
            B = C
            fB = fC
        iters += 1

    if abs(B - A) > EPSILON:
        # Fallback: solver did not converge; keep volatility unchanged and warn
        try:
            print("⚠️ glicko2 volatility solver did not converge; keeping vol unchanged.")
        except Exception:
            pass
        new_sigma = vol
    else:
        new_sigma = math.exp(A / 2.0)

    phi_star = math.sqrt(phi * phi + new_sigma * new_sigma)
    phi_prime = 1 / math.sqrt((1 / (phi_star * phi_star)) + (1 / v))
    mu_prime = mu + (phi_prime * phi_prime) * g * (score - E)

    r_prime = _to_rating(mu_prime)
    rd_prime = _to_rd(phi_prime)

    return r_prime, rd_prime, new_sigma

File: apps/api/src/lib/ratings/glicko2.ts
import { defaultRatingConfig, type RatingConfig } from "./config.js";

const TAU = 0.5;
const EPSILON = 1e-6;

function g(phi: number): number {
  return 1 / Math.sqrt(1 + (3 * phi ** 2) / Math.PI ** 2);
}

function expected(mu: number, muJ: number, phiJ: number): number {
  return 1 / (1 + Math.exp(-g(phiJ) * (mu - muJ)));
}

function toMu(rating: number): number {
  return (rating - 1500) / 173.7178;
}

function toPhi(rd: number): number {
  return rd / 173.7178;
}

function toRating(mu: number): number {
  return mu * 173.7178 + 1500;
}

function toRd(phi: number): number {
  return phi * 173.7178;
}

export function inflateRd(rd: number, days: number, config: RatingConfig = defaultRatingConfig): number {
  if (!config.g2RdIncreasePerDay) {
    return rd;
  }
  const clampedDays = Math.max(0, days);
  if (clampedDays === 0) {
    return rd;
  }
  const c = config.g2RdIncreasePerDay;
  return Math.min(Math.sqrt(rd * rd + c * c * clampedDays), config.g2DefaultRd);
}

function f(x: number, delta: number, phi: number, v: number, a: number, tau: number): number {
  const ex = Math.exp(x);
  const num = ex * (delta * delta - phi * phi - v - ex);
  const den = 2 * (phi * phi + v + ex) ** 2;
  return num / den - (x - a) / (tau * tau);
}

export type GlickoProfile = {
  rating: number;
  rd: number;
  vol: number;
  lastGameDate: string | null;
};

function daysBetween(previous: string | null, current: string): number {
  if (!previous) {
    return 0;
  }
  const previousTime = Date.parse(`${previous}T00:00:00.000Z`);
  const currentTime = Date.parse(`${current}T00:00:00.000Z`);
  if (Number.isNaN(previousTime) || Number.isNaN(currentTime)) {
    return 0;
  }
  return Math.max(0, Math.floor((currentTime - previousTime) / 86_400_000));
}

export function glicko2Update(
  player: GlickoProfile,
  opponent: GlickoProfile,
  score: number,
  matchDate: string,
  options: { tau?: number; config?: RatingConfig } = {}
): GlickoProfile {
  const tau = options.tau ?? TAU;
  const config = options.config ?? defaultRatingConfig;
  const days = daysBetween(player.lastGameDate, matchDate);
  const opponentDays = daysBetween(opponent.lastGameDate, matchDate);

  const mu = toMu(player.rating);
  const rdStar =
    days && config.g2RdIncreasePerDay
      ? Math.sqrt(player.rd * player.rd + config.g2RdIncreasePerDay * config.g2RdIncreasePerDay * days)
      : player.rd;
  const phi = toPhi(rdStar);
  const muJ = toMu(opponent.rating);
  const phiJ = toPhi(inflateRd(opponent.rd, opponentDays, config));

  const gValue = g(phiJ);
  const eValue = expected(mu, muJ, phiJ);
  const v = 1 / (gValue * gValue * eValue * (1 - eValue));
  const delta = v * gValue * (score - eValue);

  const a = Math.log(player.vol * player.vol);
  let aCandidate = a;
  let bCandidate: number;
  if (delta * delta > phi * phi + v) {
    bCandidate = Math.log(delta * delta - phi * phi - v);
  } else {
    let k = 1;
    while (f(a - k * tau, delta, phi, v, a, tau) < 0) {
      k += 1;
    }
    bCandidate = a - k * tau;
  }

  let fA = f(aCandidate, delta, phi, v, a, tau);
  let fB = f(bCandidate, delta, phi, v, a, tau);
  let iterations = 0;

  while (Math.abs(bCandidate - aCandidate) > EPSILON && iterations < 60) {
    const cCandidate = aCandidate + ((aCandidate - bCandidate) * fA) / (fB - fA);
    const fC = f(cCandidate, delta, phi, v, a, tau);
    if (fC * fB < 0) {
      aCandidate = bCandidate;
      fA = fB;
      bCandidate = cCandidate;
      fB = fC;
    } else {
      fA /= 2;
      bCandidate = cCandidate;
      fB = fC;
    }
    iterations += 1;
  }

  const newSigma = Math.abs(bCandidate - aCandidate) > EPSILON ? player.vol : Math.exp(aCandidate / 2);
  const phiStar = Math.sqrt(phi * phi + newSigma * newSigma);
  const phiPrime = 1 / Math.sqrt(1 / (phiStar * phiStar) + 1 / v);
  const muPrime = mu + phiPrime * phiPrime * gValue * (score - eValue);

  return {
    rating: toRating(muPrime),
    rd: toRd(phiPrime),
    vol: newSigma,
    lastGameDate: matchDate
  };
}
